File: app/src/main/java/com/usher/demo/utils/PermissionUtil.java
package com.usher.demo.utils;

import android.app.Activity;
import android.content.Context;
import android.content.pm.PackageManager;

import java.util.ArrayList;
import java.util.List;

import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;

public class PermissionUtil {
    private static PublishSubject<Boolean> subject = PublishSubject.create();

    public static Observable<Boolean> requestPermission(Context context, String permission) {
        return requestPermissions(context, new String[]{permission});
    }

    public static Observable<Boolean> requestPermissions(Context context, String[] permissions) {
        List<String> permissionList = new ArrayList<>();
        for (String permission : permissions) {
            if (PackageManager.PERMISSION_GRANTED != ContextCompat.checkSelfPermission(context, permission)) {
                permissionList.add(permission);
            }
        }

        if (!permissionList.isEmpty()) {
            ActivityCompat.requestPermissions((Activity) context, permissionList.toArray(new String[permissionList.size()]), 0);
        } else {
            return Observable.just(true);
        }

        return subject;
    }

    public static void grant() {
        subject.onNext(true);
    }

    public static void deny() {
        subject.onNext(false);
    }
}

File: TwigCodesUI/src/main/java/com/twigcodes/ui/util/PermissionUtil.kt
package com.twigcodes.ui.util

import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject

object PermissionUtil {
    private val mPermissionSubject = PublishSubject.create<Boolean>()

    fun requestPermission(context: Context, permission: String): Observable<Boolean> = requestPermissions(context, arrayOf(permission))

    fun requestPermissions(context: Context, permissions: Array<String>): Observable<Boolean> {
        val permissionList = permissions.filter { PackageManager.PERMISSION_DENIED == ContextCompat.checkSelfPermission(context, it) }

        return if (permissionList.isNotEmpty()) {
            ActivityCompat.requestPermissions(context as Activity, permissionList.toTypedArray(), 0)
            mPermissionSubject
        } else {
            Observable.just(true)
        }
    }

    fun grant() = mPermissionSubject.onNext(true)

    fun deny() = mPermissionSubject.onNext(false)
}